pie/models/decoder.py

```diff
@@ -328,35 +328,100 @@ def predict_max(self, enc_outs, lengths,
-        hyps, scores = [], 0
+        hyps, scores = [], [0 for _ in range(batch)]
+
+        # As we go, we'll reduce the tensor size by popping finished prediction
+        #  To keep adding new characters to the right words, we
+        #  store and keep updated a table of {Tensor Row ID: Batch Original ID}
+        #  where Batch Original ID is the Word ID (batch_size = number of words)
+        tensor_to_batch_indexes = {
+            x: x for x in range(batch)
+        }
 
         for _ in range(max_seq_len):
-            if mask.sum().item() == 0:
-                break
 
-            # prepare input
-            emb = self.embs(inp)
+            # Prepare input
+            #    Context is NEVER changed after the method has been called
+            emb = self.embs(inp)  # Tensor(batch_size x emb_size)
             if context is not None:
-                emb = torch.cat([emb, context], dim=1)
-            # run rnn
-            emb = emb.unsqueeze(0)
+                emb = torch.cat([emb, context], dim=1)  # Tensor(batch_size x (emb_size+context_size))
+
+            # Run rnn
+            emb = emb.unsqueeze(0)  # Tensor(1 x batch_size x emb size(+context))
+
+            # hidden is gonna be reused by the next iteration
+            #   outs is specific to the current run
             outs, hidden = self.rnn(emb, hidden)
+            # Hidden : Tensor(1 x batch_size x emb_size)
+
             outs, _ = self.attn(outs, enc_outs, lengths)
             outs = self.proj(outs).squeeze(0)
-            # get logits
+
+            # Get logits
             probs = F.log_softmax(outs, dim=1)
-            # sample and accumulate
-            score, inp = probs.max(1)
-            hyps.append(inp.tolist())
-            mask = mask * (inp != eos).long()
-            score = score.cpu()
-            score[mask == 0] = 0
-            scores += score
+
+            # Sample and accumulate
+            #  Score are the probabilities
+            #  Inp are the new characters (as int) we are adding to our predictions
+            score, inp = probs.max(1)  # (Tensor(batch_size, dtype=float), Tensor(batch_size, dtype=int))
+
+            # We create a mask of value that are not ending the string
+            non_eos = (inp != eos)  # Tensor(batch_size, dtype=bool)
+
+            # Using this mask, we retrieve the Indexes of items that are not EOS
+            #  nonzero() returns a tuple where the first item is a Tensor
+            #  with the indexes. It can be use as a selector for other tensors (see below)
+            keep, *_ = torch.nonzero(non_eos, as_tuple=True)  # Tensor(dtype=int)
+
+            # Add new chars to hypotheses
+            #   We prepare a list the size of the output, filling it with EOS
+            #   We then iterate over score and inp (same sized tensors) and add to our
+            prediction_run_output = [eos for _ in range(batch)]
+
+            for ind, (hyp, sc) in enumerate(zip(inp.tolist(), score.tolist())):
```

Comment:
I find this bit particularly confusing and I think there is no need for a for loop here, which goes against the tensor-level operations used in the rest of the package.

I would rewrite all this using a mask. Check the following points:
- the keys in tensor_to_batch_indices are always 0 to number of remaining hypothesis. so the use of a dictionary is redundant. You could keep a tensor with the target indices.
- prediction_run_output (btw. I cannot imagine what the variable represents on the basis of the name only, perhaps look for a more informative name) can be updated at once using masks (same for scores)

As a strategy you could define an output tensor at once of size (batch x max_length) for output hyps and index it directly after each step. For the output scores, you can keep a 1D tensor and update it using a mask that indexes to words not yet finished. When the process is done you can rearrange hyps into the expected format.

Reply:
I do not completely get what you mean (unfortunately) and I am not sure this is gonna make the code clearer. I'd argue it will make it less clear actually. You're the code owner, but outside the variable name, I really feel like the code is accessible and easy to tweak for outsider as is. And we have loops here and there (for example beams) even if these might be different.

Reply:
Done